### PR TITLE
Improve perf of avg_pool2d by rewriting as a reduction

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116756

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -4276,60 +4269,53 @@ def _avg_poolnd(
...
+        def ones_loader(idx):
+            prefix = idx[:-dim]
+            bh = idx[-dim:]
+            lower = [-padding[i] if count_include_pad else 0 for i in range(dim)]
+            upper = [
+                h[i] + padding[i] if count_include_pad else h[i] for i in range(dim)
+            ]
+            masks = [range_mask(bh[i], upper[i], lower[i]) for i in range(dim)]
+            mask = functools.reduce(ops.and_, masks)
+            return ops.masked(mask, lambda: ops.constant(1, dtype), 0)
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:4313]:
```suggestion
            return ops.to_dtype(mask, dtype)
```

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -4276,60 +4269,53 @@ def _avg_poolnd(
...
-    rv = Pointwise.create(
-        device=x.get_device(),
-        dtype=dtype,
-        inner_fn=fn,
-        ranges=new_size,
-    )
-    # TODO(jansel): should we force these to be realized?
-    return rv
+        # TODO(jansel): optimize to do `int(x<h)` rather than `x<h?1:0`
+        result = div(fn_sum(x_loader), fn_sum(ones_loader))
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:4316]: For large kernel sizes I expect `sum(ones_loader)` is quite inefficient. Can we write it as an index_expr instead? Something like using min and max to get the in-bounds shape in each dim and multiply them together to get the volume.

**Review by peterbell10 — COMMENTED** [review]
Could do with some benchmarks

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (isuruf) [reply torch/_inductor/lowering.py:4293]: I did write such a function and the runtime came out to be slower than this version.

Comment (peterbell10) [reply torch/_inductor/lowering.py:4293]: Do you still have the code? I'd be curious to see why it was worse.

Comment (isuruf) [reply torch/_inductor/lowering.py]: I was fiddling around with reduction hints at the same time and `sum(ones_loader)` was better on one of those hints, but that hint was not good anyway. I've pushed a commit.

**Comment by isuruf** [comment]
Here are some results

Benchmark                                                                                          | main       | PR 1st commit | PR last commit | Eager
------------------------------------------------------------------------------------------|-------------|----------|----------|------
{'kernel_size': 3, 'stride': 2, 'padding': 1, 'shape': (4, 128, 56, 56)}       | 31.7       | 36       | 31       | 46.3
{'kernel_size': 3, 'stride': 3, 'padding': 1, 'shape': (4, 128, 56, 56)}       | 29.7       | 28       | 26       | 35.0
{'kernel_size': 3, 'stride': 3, 'padding': 0, 'shape': (4, 128, 56, 56)}       | 25.7       | 25       | 25       | 34.1
{'kernel_size': 3, 'stride': 2, 'padding': 0, 'shape': (4, 128, 56, 56)}       | 28.8       | 29       | 29       | 44.5
{'kernel_size': 9, 'stride': 4, 'padding': 1, 'shape': (4, 128, 56, 56)}       | 60.7       | 77       | 72       | 63.1
{'kernel_size': 9, 'stride': 9, 'padding': 1, 'shape': (4, 128, 56, 56)}       | 27.9       | 21       | 21       | 30.7
{'kernel_size': 9, 'stride': 9, 'padding': 0, 'shape': (4, 128, 56, 56)}       | 27.2       | 21       | 21       | 29.8
{'kernel_size': 9, 'stride': 4, 'padding': 0, 'shape': (4, 128, 56, 56)}       | 59.2       | 50       | 50       | 62.1

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -4290,7 +4290,36 @@ def ones_loader(idx):
...
+                )
+                hend = ops.minimum(
+                    ops.add(hstart, kernel_h),
+                    ops.add(ops.index_expr(h[i], torch.int32), pad_h),
+                )
+                if not count_include_pad:
+                    hstart = ops.maximum(hstart, ops.constant(0, torch.int32))
+                    hend = ops.minimum(hend, ops.index_expr(h[i], torch.int32))
+                divide_factors.append(ops.sub(hend, hstart))
+            return functools.reduce(ops.mul, divide_factors)
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:4313]: This should be all done in sympy and leave the `ops.index_expr` call until the very end.

**Comment by isuruf** [comment]
@peterbell10, there's only one slowdown as seen in the table above and that can be fixed by changing https://github.com/pytorch/pytorch/blob/6978c3ddf3fe292f7cd91f45b4815aadc0c30e91/torch/_inductor/triton_heuristics.py#L1247 to 
```
triton_config_reduction(size_hints, 64, 1, num_warps=8), 
```
or
```
triton_config_reduction(size_hints, 64, 2, num_warps=8), 
```

Any ideas on how to do that without making other kernels perform slower?

**Comment by peterbell10** [comment]
`8 wraps * 32 threads per warp = 256 threads` yet your proposed configs only have 64 and 128 elements of data respectively. So neither of these configs really make sense.

Also I'm surprised you're even hitting the `reduction` heuristics for the kernel sizes you show. I would expect those to generate a `persistent_reduction`. Could you share some of the generated code?

**Comment by peterbell10** [comment]
If those configs are being used though, that would mean it's doing the reduction entirely within a single thread which would be equivalent to the unrolled reduction just without actually unrolling the loop.

**Comment by isuruf** [comment]
Benchmark script:

<details>

```python
import torch
from torch.testing import make_tensor
from torch.fx.experimental.proxy_tensor import make_fx
from torch.utils.benchmark import Timer, Compare
from torch._inductor.compile_fx import compile_fx_inner, cudagraphify_impl
from torch._inductor.decomposition import decompositions
from itertools import product
from functools import partial

aten = torch.ops.aten

torch._logging.set_logs(output_code=True)

benchmark_name = "avg_pool"
Ss = [
    [3, 5, 1000000, 10, 10, 0],
    [3, 5, 1000, 1000, 4, 4, 0],
    [3, 5, 100, 100, 100, 2, 2, 0],
    [3, 5, 1000000, 10, 10, 1],
    [3, 5, 1000, 1000, 4, 4, 1],
    [3, 5, 100, 100, 100, 2, 2, 1],
    [3, 5, 1000000, 10, 1, 1],
    [3, 5, 1000, 1000, 4, 1, 1],
    [3, 5, 100, 100, 100, 2, 1, 1],
]
Ss = [
    #[4, 128, 56, 56, 3, 2, 1],
    #[4, 128, 56, 56, 3, 3, 1],
    #[4, 128, 56, 56, 3, 3, 0],
    #[4, 128, 56, 56, 3, 2, 0],
    [4, 128, 56, 56, 9, 4, 1],
    #[4, 128, 56, 56, 9, 9, 1],
    #[4, 128, 56, 56, 9, 9, 0],
    #[4, 128, 56, 56, 9, 4, 0],
]


def gen_inputs():
    for inp in Ss:
        shape = inp[:-3]
        yield [torch.randn(shape, dtype=torch.float32, device="cuda"), inp[-3:]]


def benchmark(label, f, x, kwargs):
    kwargs = kwargs.copy()
    kwargs["shape"] = tuple(x.shape)
    return Timer(
        "f([x,])",
        globals=locals(),
        label=benchmark_name,
        description=label,
        sub_label=str(kwargs),
        num_threads=torch.get_num_threads(),
    ).blocked_autorange(min_run_time=4)


def compare(args):
    x, props = args
    print(f"{tuple(x.shape)}")
    if x.dim() == 3:
        aten_func = aten.avg_pool1d
    elif x.dim() == 4:
        aten_func = aten.avg_pool2d
    elif x.dim() == 5:
        aten_func = aten.avg_pool3d

    kwargs = {
        "kernel_size": props[0],
        "stride": props[1],
        "padding": props[2],
        "count_include_pad": False,
    }

    def f(args):
        (x,) = args
        val = aten_func(x, **kwargs)
        return (val,)

    args = [x]

    # decomposed = make_fx(f, decomposition_table=decompositions, tracing_mode="fake")(args)
    # compiled_decomposed = compile_fx_inner(decomposed, args, cudagraphs=False)
    # yield benchmark(f"Decomposed", compiled_decomposed, *args, kwargs)

    non_decomposed = make_fx(f, tracing_mode="fake")(args)
    compiled_nondecomposed = compile_fx_inner(non_decomposed, args, cudagraphs=False)
    yield benchmark("Lowering", compiled_nondecomposed, *args, kwargs)
    # Just show the first two generated kernels
    torch._logging.set_logs(output_code=False)

    cuda_f = cudagraphify_impl(f, args, static_input_idxs=tuple(range(len(args))))
    yield benchmark(f"Eager", cuda_f, *args, kwargs)


results = []
for x in gen_inputs():
    for res in compare(x):
        results.append(res)

compare = Compare(results)
compare.trim_significant_figures()
compare.print()
```

</details>

Here's the generated code

<details>

```
from ctypes import c_void_p, c_long
import torch
import math
import random
import os
import tempfile
from math import inf, nan
from torch._inductor.hooks import run_intermediate_hooks
from torch._inductor.utils import maybe_profile
from torch._inductor.codegen.memory_planning import _align as align

from torch import device, empty, empty_strided
from torch._inductor.codecache import AsyncCompile
from torch._inductor.select_algorithm import extern_kernels
from torch._inductor.codegen.multi_kernel import MultiKernelCall

aten = torch.ops.aten
inductor_ops = torch.ops.inductor
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
alloc_from_pool = torch.ops.inductor._alloc_from_pool
reinterpret_tensor = torch.ops.inductor._reinterpret_tensor
async_compile = AsyncCompile()


# kernel path: /tmp/torchinductor_isuruf/uw/cuwyta6ldkwjmaw5xakohycmrv2d3rruvws4gnwf4fiwmnrxggvz.py
# Source Nodes: [], Original ATen: []

triton_red_fused_0 = async_compile.triton('triton_', '''
import triton
import triton.language as tl
from torch._inductor.ir import ReductionHint
from torch._inductor.ir import TileHint
from torch._inductor.triton_heuristics import AutotuneHint, reduction
from torch._inductor.utils import instance_descriptor
from torch._inductor import triton_helpers

@reduction(
    size_hints=[131072, 128],
    reduction_hint=ReductionHint.DEFAULT,
    filename=__file__,
    triton_meta={'signature': {0: '*fp32', 1: '*fp32', 2: 'i32', 3: 'i32'}, 'device': 0, 'device_type': 'cuda', 'constants': {}, 'configs': [instance_descriptor(divisible_by_16=(0, 1, 2), equal_to_1=(), ids_of_folded_args=(), divisible_by_8=(2,))]},
    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_red_fused_0', 'mutated_arg_names': ['in_out_ptr0'], 'no_x_dim': False}
)
@triton.jit
def triton_(in_out_ptr0, in_ptr0, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
    xnumel = 86528
    rnumel = 81
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = xindex < xnumel
    rbase = tl.arange(0, RBLOCK)[None, :]
    x1 = (xindex // 13) % 13
    x0 = xindex % 13
    x2 = (xindex // 169)
    _tmp15 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
    x6 = xindex
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r4 = (rindex // 9)
        r3 = rindex % 9
        tmp0 = (-1) + r4 + (4*x1)
        tmp1 = tl.full([1, 1], 0, tl.int64)
        tmp2 = tmp0 >= tmp1
        tmp3 = tl.full([1, 1], 56, tl.int64)
        tmp4 = tmp0 < tmp3
        tmp5 = tmp2 & tmp4
        tmp6 = (-1) + r3 + (4*x0)
        tmp7 = tmp6 >= tmp1
        tmp8 = tmp6 < tmp3
        tmp9 = tmp7 & tmp8
        tmp10 = tmp5 & tmp9
        tmp11 = tl.load(in_ptr0 + ((-57) + r3 + (4*x0) + (56*r4) + (224*x1) + (3136*x2)), rmask & tmp10 & xmask, eviction_policy='evict_first', other=0.0)
        tmp12 = tl.full(tmp11.shape, 0.0, tmp11.dtype)
        tmp13 = tl.where(tmp10, tmp11, tmp12)
        tmp14 = tl.broadcast_to(tmp13, [XBLOCK, RBLOCK])
        tmp16 = _tmp15 + tmp14
        _tmp15 = tl.where(rmask & xmask, tmp16, _tmp15)
    tmp15 = tl.sum(_tmp15, 1)[:, None]
    tmp17 = ((tl.math.max(0, (-1) + (4*x0)))*(tl.math.max(0, (-1) + (4*x1)))) + ((tl.math.min(56, 8 + (4*x0)))*(tl.math.min(56, 8 + (4*x1)))) + ((-1)*(tl.math.max(0, (-1) + (4*x0)))*(tl.math.min(56, 8 + (4*x1)))) + ((-1)*(tl.math.max(0, (-1) + (4*x1)))*(tl.math.min(56, 8 + (4*x0))))
    tmp18 = tmp15 / tmp17
    tl.debug_barrier()
    tl.store(in_out_ptr0 + (x6), tmp18, xmask)
''')

import triton
import triton.language as tl
from torch._inductor.triton_heuristics import grid, start_graph, end_graph
from torch._C import _cuda_getCurrentRawStream as get_raw_stream


async_compile.wait(globals())
del async_compile

def call(args):
    args_1, = args
    args.clear()
    assert_size_stride(args_1, (4, 128, 56, 56), (401408, 3136, 56, 1))
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0)
        buf0 = empty((4, 128, 13, 13), device='cuda', dtype=torch.float32)
        buf1 = buf0; del buf0  # reuse
        # Source Nodes: [], Original ATen: []
        stream0 = get_raw_stream(0)
        triton_red_fused_0.run(buf1, args_1, 86528, 81, grid=grid(86528), stream=stream0)
        del args_1
        return (buf1, )


def benchmark_compiled_module(times=10, repeat=10):
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    args_1 = rand_strided((4, 128, 56, 56), (401408, 3136, 56, 1), device='cuda:0', dtype=torch.float32)
    fn = lambda: call([args_1])
    return print_performance(fn, times=times, repeat=repeat)


if __name__ == "__main__":
    from torch._inductor.wrapper_benchmark import compiled_module_main
    compiled_module_main('None', benchmark_compiled_module)
```

</details>

Note that the there's a difference in a GeForce RTX 2060 only. On Quadro RTX 8000, the patch doesn't do anything and this PR is strictly better than the eager mode.

**Comment by isuruf** [comment]
As discussed, this PR is slower only for the older GPU GeForce RTX 2060 and is faster for newer GPUs and we only care about the newer GPUs.

Ready for a final review.

File: torch/_inductor/lowering.py [hunk 4]
```diff
@@ -4386,66 +4384,67 @@ def _avg_poolnd(
...
+                divide_factors.append(factor)
+            return functools.reduce(ops.mul, divide_factors)
+
+        divide_factor = Pointwise.create(
+            device=x.get_device(),
+            dtype=dtype,
+            inner_fn=fn_count,
+            ranges=new_size,
+        )
+        result = div(rv, divide_factor)
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:4445]: Benchmarks show a real slowdown, I confirmed that the `adv_inception_v3` slowdown is reproducible and really calls `avg_pool2d`.

https://hud.pytorch.org/benchmark/timm_models/inductor_with_cudagraphs?startTime=Sun,%2004%20Feb%202024%2013:25:02%20GMT&stopTime=Sun,%2011%20Feb%202024%2013:25:02%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=gh/isuruf/17/head&lCommit=f610d4dc4de17fb89b9950d05c5f39a9e4ed2019&rBranch=main&rCommit=5f6b35915af98fac741998edb55aeeba4aaa3794

One interesting note is that this division is not being fused with the reduction kernel. The cause is a known issue where reductions iterate in a different order which ends up blocking subsequent pointwise fusions in the scheduler.  

Here the reduciton loop has size
```
buf191.sizes = ([128, 192, 35, 35], [3, 3])
```
but the pointwise op has
```
buf192.sizes = ([128, 35, 35, 192], [])
```
The reordering of the 192-length dimension is what blocks the fusion in this case. Should be fixed by #100331 but that's still a work in progress.

Comment (lezcano) [reply torch/_inductor/lowering.py:4445]: @shunting314 a real world example where https://github.com/pytorch/pytorch/pull/100331 would help.

File: torch/_inductor/lowering.py [hunk 5]
```diff
@@ -4386,66 +4385,79 @@ def _avg_poolnd(
...
-            if total is None:
-                total = val
-            else:
-                total = ops.add(val, total)
-        return total
+    # TODO: remove this when #100331 is merged. We only do this
+    # for window_size <=25 to avoid performance regressions compared
+    # to the previous algorithm which unrolled manually for <=25
+    context = (
+        config.patch(unroll_reductions_threshold=25)
```
Comment (lezcano) [inline torch/_inductor/lowering.py:4408]: The bad thing about this patch is that it stops us from writing all these as decomps, as we can't perform this trick (or the realize trick below) at a decomp level :/

Comment (isuruf) [reply torch/_inductor/lowering.py:4408]: I have a workaround at the end of https://github.com/pytorch/pytorch/commit/b8ad88b291c22b04b3679920d198bbb91179774f. This makes the decomp as fast as the lowering here.

Comment (isuruf) [reply torch/_inductor/lowering.py:4408]: The workaround may not work all the time. Let me try to fix that.

Comment (isuruf) [reply torch/_inductor/lowering.py:4408]: Yes, you are right. We can't affect configs from decompositions.

**Comment by isuruf** [comment]
Benchmark timings

<details>

```
[-------------------------------------------------------- avg_pool -------------------------------------------------------------------------------------------------]
                                                                                                           |  eager  | main     |   pr1    |   pr2    |    pr3 
12 threads: ---------------------------------------------------------------------------------------------------------------------------------------------------------
{'kernel_size': 10, 'stride': 10, 'padding': 0, 'count_include_pad': False, 'shape': (3, 5, 1000000)}      |    244  |     112  |     113  |     112  |     113
{'kernel_size': 4, 'stride': 4, 'padding': 0, 'count_include_pad': False, 'shape': (3, 5, 1000, 1000)}     |    188  |     142  |     142  |     142  |     142
{'kernel_size': 2, 'stride': 2, 'padding': 0, 'count_include_pad': False, 'shape': (3, 5, 100, 100, 100)}  |    346  |     337  |      98  |      97  |      98
{'kernel_size': 10, 'stride': 10, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 1000000)}      |    230  |     115  |     115  |     109  |     109
{'kernel_size': 4, 'stride': 4, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 1000, 1000)}     |    225  |     151  |     151  |     159  |     159
{'kernel_size': 2, 'stride': 2, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 100, 100, 100)}  |    354  |     345  |     118  |     107  |     107
{'kernel_size': 10, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 1000000)}       |   1670  |     724  |     724  |     605  |     605
{'kernel_size': 4, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 1000, 1000)}     |   2578  |    1442  |    1442  |    1027  |    1028
{'kernel_size': 3, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 100, 100, 100)}  |   4360  |    4350  |    4350  |    4350  |    4330
{'kernel_size': 2, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 100, 100, 100)}  |   2549  |    2539  |     755  |     701  |     701
{'kernel_size': 3, 'stride': 2, 'padding': 1, 'count_include_pad': False, 'shape': (4, 128, 56, 56)}       |     69  |      30  |      30  |      30  |      30
{'kernel_size': 3, 'stride': 3, 'padding': 1, 'count_include_pad': False, 'shape': (4, 128, 56, 56)}       |     42  |      17  |      18  |      18  |      18
{'kernel_size': 3, 'stride': 3, 'padding': 0, 'count_include_pad': False, 'shape': (4, 128, 56, 56)}       |     42  |      15  |      16  |      15  |      16
{'kernel_size': 3, 'stride': 2, 'padding': 0, 'count_include_pad': False, 'shape': (4, 128, 56, 56)}       |     68  |      22  |      22  |      22  |      22
{'kernel_size': 3, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (128, 35, 35, 192)}     |   3693  |    1034  |    1034  |     921  |     921
{'kernel_size': 9, 'stride': 4, 'padding': 1, 'count_include_pad': False, 'shape': (4, 128, 56, 56)}       |     76  |      66  |      67  |      66  |      74
{'kernel_size': 9, 'stride': 9, 'padding': 1, 'count_include_pad': False, 'shape': (4, 128, 56, 56)}       |     85  |      76  |      76  |      76  |      23
{'kernel_size': 9, 'stride': 9, 'padding': 0, 'count_include_pad': False, 'shape': (4, 128, 56, 56)}       |     70  |      61  |      61  |      61  |      22
{'kernel_size': 9, 'stride': 4, 'padding': 0, 'count_include_pad': False, 'shape': (4, 128, 56, 56)}       |     76  |      67  |      67  |      67  |      49
{'kernel_size': 5, 'stride': 5, 'padding': 0, 'count_include_pad': False, 'shape': (4, 3, 1000, 1000)}     |    131  |      73  |      74  |      73  |      72

```

**Comment by lezcano** [comment]
This looks very good. @vfdev-5 could you help double check the benchmarks / provide a set of relevant inputs shapes for us to benchmark?

**Comment by peterbell10** [comment]
I've kicked off torchbench for this PR and the previous one.

**Comment by peterbell10** [comment]
The new hud results look pretty neutral which isn't too surprising as I don't think it hits larger kernel size. In that case I think lets just get some microbenchmarks for half precision.

**Comment by isuruf** [comment]
Benchmarks for bfloat16
```
[-------------------------------------------------------------------------- avg_pool -----------------------------------------------------------------------------------]
                                                                                                                                          |  Lowering  |  Eager   | PR
12 threads: -------------------------------------------------------------------------------------------------------------------------------------------------------------
      {'kernel_size': 10, 'stride': 10, 'padding': 0, 'count_include_pad': False, 'shape': (3, 5, 1000000), 'dtype': torch.bfloat16}      |     20     |    49    |  21
      {'kernel_size': 4, 'stride': 4, 'padding': 0, 'count_include_pad': False, 'shape': (3, 5, 1000, 1000), 'dtype': torch.bfloat16}     |     19     |    52    |  20
      {'kernel_size': 2, 'stride': 2, 'padding': 0, 'count_include_pad': False, 'shape': (3, 5, 100, 100, 100), 'dtype': torch.bfloat16}  |     24     |    89    |  24
      {'kernel_size': 10, 'stride': 10, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 1000000), 'dtype': torch.bfloat16}      |     20     |    49    |  21
      {'kernel_size': 4, 'stride': 4, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 1000, 1000), 'dtype': torch.bfloat16}     |     20     |    62    |  20
      {'kernel_size': 2, 'stride': 2, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 100, 100, 100), 'dtype': torch.bfloat16}  |     31     |    91    |  31
      {'kernel_size': 10, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 1000000), 'dtype': torch.bfloat16}       |    141     |   393    | 141
      {'kernel_size': 4, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 1000, 1000), 'dtype': torch.bfloat16}     |    241     |   576    | 241
      {'kernel_size': 3, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 100, 100, 100), 'dtype': torch.bfloat16}  |    937     |   941    | 916
      {'kernel_size': 2, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (3, 5, 100, 100, 100), 'dtype': torch.bfloat16}  |    158     |   548    | 158
      {'kernel_size': 3, 'stride': 2, 'padding': 1, 'count_include_pad': False, 'shape': (4, 128, 56, 56), 'dtype': torch.bfloat16}       |     16     |    18    |  16
      {'kernel_size': 3, 'stride': 3, 'padding': 1, 'count_include_pad': False, 'shape': (4, 128, 56, 56), 'dtype': torch.bfloat16}       |     16     |    15    |  16
      {'kernel_size': 3, 'stride': 3, 'padding': 0, 'count_include_pad': False, 'shape': (4, 128, 56, 56), 'dtype': torch.bfloat16}       |     16     |    15    |  16
      {'kernel_size': 3, 'stride': 2, 'padding': 0, 'count_include_pad': False, 'shape': (4, 128, 56, 56), 'dtype': torch.bfloat16}       |     16     |    18    |  16
      {'kernel_size': 3, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (128, 35, 35, 192), 'dtype': torch.bfloat16}     |    182     |   909    | 182
      {'kernel_size': 9, 'stride': 4, 'padding': 1, 'count_include_pad': False, 'shape': (4, 128, 56, 56), 'dtype': torch.bfloat16}       |     15     |    15    |  18
      {'kernel_size': 9, 'stride': 9, 'padding': 1, 'count_include_pad': False, 'shape': (4, 128, 56, 56), 'dtype': torch.bfloat16}       |     15     |    19    |  16
      {'kernel_size': 9, 'stride': 9, 'padding': 0, 'count_include_pad': False, 'shape': (4, 128, 56, 56), 'dtype': torch.bfloat16}       |     15     |    19    |  16
      {'kernel_size': 9, 'stride': 4, 'padding': 0, 'count_include_pad': False, 'shape': (4, 128, 56, 56), 'dtype': torch.bfloat16}       |     15     |    16    |  16
      {'kernel_size': 5, 'stride': 5, 'padding': 0, 'count_include_pad': False, 'shape': (4, 3, 1000, 1000), 'dtype': torch.bfloat16}     |     16     |    30    |  17
```

**Comment by peterbell10** [comment]
I think you need larger inputs because timings around ~15 us usually mean the kernel is overhead bound. To save time you can focus on kernels with >= 25 elements so we're exercising the reduction kernel path. It would also be good to see a `stride=1` reduction kernel example.

**Comment by isuruf** [comment]
Hmm, the benchmarks now looks really bad
```
[-------------------------------------------------------------------------- avg_pool -------------------------------------------------------------------------]
                                                                                                                                          |  Lowering  |  Eager
12 threads: ---------------------------------------------------------------------------------------------------------------------------------------------------
      {'kernel_size': 9, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.float32}    |    25230   |   7012
      {'kernel_size': 9, 'stride': 1, 'padding': 0, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.float32}    |    16900   |   6941
      {'kernel_size': 9, 'stride': 4, 'padding': 1, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.float32}    |     1049   |    681
      {'kernel_size': 9, 'stride': 9, 'padding': 1, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.float32}    |      246   |    269
      {'kernel_size': 9, 'stride': 9, 'padding': 0, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.float32}    |      246   |    269
      {'kernel_size': 9, 'stride': 4, 'padding': 0, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.float32}    |      639   |    671
      {'kernel_size': 25, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.float32}   |   155600   |  22770
      {'kernel_size': 25, 'stride': 1, 'padding': 0, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.float32}   |   100700   |  22820
      {'kernel_size': 25, 'stride': 4, 'padding': 1, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.float32}   |     8200   |   3444
      {'kernel_size': 25, 'stride': 9, 'padding': 1, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.float32}   |     1170   |    811
      {'kernel_size': 25, 'stride': 9, 'padding': 0, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.float32}   |      886   |    807
      {'kernel_size': 25, 'stride': 4, 'padding': 0, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.float32}   |     6313   |   3373
      {'kernel_size': 9, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.bfloat16}   |    25970   |   6722
      {'kernel_size': 9, 'stride': 1, 'padding': 0, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.bfloat16}   |    17180   |   6718
      {'kernel_size': 9, 'stride': 4, 'padding': 1, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.bfloat16}   |     1045   |    524
      {'kernel_size': 9, 'stride': 9, 'padding': 1, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.bfloat16}   |      237   |    175
      {'kernel_size': 9, 'stride': 9, 'padding': 0, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.bfloat16}   |      148   |    174
      {'kernel_size': 9, 'stride': 4, 'padding': 0, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.bfloat16}   |      622   |    521
      {'kernel_size': 25, 'stride': 1, 'padding': 1, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.bfloat16}  |   145600   |  22120
      {'kernel_size': 25, 'stride': 1, 'padding': 0, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.bfloat16}  |   104200   |  22100
      {'kernel_size': 25, 'stride': 4, 'padding': 1, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.bfloat16}  |     8388   |   1993
      {'kernel_size': 25, 'stride': 9, 'padding': 1, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.bfloat16}  |     1108   |    472
      {'kernel_size': 25, 'stride': 9, 'padding': 0, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.bfloat16}  |      826   |    471
      {'kernel_size': 25, 'stride': 4, 'padding': 0, 'count_include_pad': False, 'shape': (10, 10, 1000, 1000), 'dtype': torch.bfloat16}  |     6526   |   1975
```

**Comment by peterbell10** [comment]
Ouch, yeah that's pretty bad. It may be worth investigating the performance issues a bit. Given that these large kernel sizes don't appear in the benchmark suite though, I'm inclined to say don't bother. 

Perhaps @vfdev-5 could comment though, are these large kernel sizes used in any models you're aware of? (Large here is >= 25 elements).

**Comment by isuruf** [comment]
Going to close this as I don't see a good way forward.

**Comment by eellison** [comment]
@isuruf I just ran into an example of us not emitting max_pool2d because of window_size being too large. I wonder if large window would be a good case for us emitting avg/max pool as a reduction.

See https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmprkbp2X/-_0_1_0/inductor_output_code_ck2t6p46e7xnymuv5hphu77hwcwcw4q4s5mlwupvwxfeet2u7d5n_22.txt?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=100

```
        # Topologically Sorted Source Nodes: [input_165, x_24], Original ATen: [aten.leaky_relu, aten.max_pool2d_with_indices]
        buf197 = torch.ops.aten.max_pool2d_with_indices.default(buf196, [9, 9], [1, 1], [4, 4])
        buf198 = buf197[0]
        assert_size_stride(buf198, (8, 512, 12, 16), (98304, 192, 16, 1))
        del buf197
        # Topologically Sorted Source Nodes: [x_25], Original ATen: [aten.max_pool2d_with_indices]
        buf200 = torch.ops.aten.max_pool2d_with_indices.default(buf196, [13, 13], [1, 1], [6, 6])
        buf201 = buf200[0]
        assert_size_stride(buf201, (8, 512, 12, 16), (98304, 192, 16, 1))
```

**Pull request closed without merging**